crates/services/p2p/src/codecs/bounded.rs

```diff
@@ -0,0 +1,156 @@
+use std::{
+    io,
+    marker::PhantomData,
+};
+
+use async_trait::async_trait;
+use futures::{
+    AsyncRead,
+    AsyncReadExt,
+    AsyncWriteExt,
+};
+use libp2p::request_response;
+use strum::IntoEnumIterator as _;
+
+use crate::request_response::{
+    messages::{
+        RequestMessage,
+        V1ResponseMessage,
+        V2ResponseMessage,
+    },
+    protocols::RequestResponseProtocol,
+};
+
+use super::{
+    DataFormatCodec,
+    RequestResponseProtocols,
+};
+
+#[derive(Debug, Clone)]
+pub struct BoundedCodec<Format> {
+    _format: PhantomData<Format>,
```

Comment:
Following my other suggestion I'd suggest skipping the `PhantomData` here. If the `Format` is an empty struct, there's no cost to holding it directly.

```suggestion
    _format: Format,
```